README.md
# sumologic-client

A JavaScript client for Sumo Logic APIs.

## Requirements

This library requires **NodeJS 10+** as it uses async iterators to stream chunks of data from Sumo Logic backend (see example below).

## Features

Only Search Job API is supported at the moment. Supported features:

-   Streams search results as they are gathered by Sumo Logic backend.
-   Notifies you of state changes on the backend ("Search started" → "Gathering results" → "Finished").
-   Handles "too many requests", and other errors on the backend.
-   Cleans up properly after search is done.

Not supported:

-  Aggregations
-  Metrics

## Install

```sh
$ npm install @moneytree/sumologic-client
```

## Prerequisites

You will need to obtain Sumo Logic access keys to use this library. Both user level and admin level access keys will work.

To create a user level access keys, go to your profile settings in Sumo Logic dashboard, and create a key in the *My Access Keys* section.

## Sample Use

```javascript
const { Search } = require('sumologic-client');
(async () => {
  const client = new Search({
    // client config (see below)
  });

  const searchParams = {
    query: '<sumologic_query>',
    // dates are ISO formatted
    from: '2019-06-25T10:14:31+09:00',
    to: '2019-06-25T17:14:31+09:00'
  };

  const it = await client.getMessageIterator(searchParams);

  for await (let response of it) {
      // check the latest backend state and total
      // number of results gathered so far:
      console.log(response.state);

      // new search results since the last iteration:
      console.log(response.results);
  }
})();
```

The config object passed to constructor can have the following options:

```js
{
  // [required] Your access ID
  accessId: '*******',

  // [required] Your access key
  accessKey: '*******',

  // [required] API endpoint as described here:
  // https://help.sumologic.com/APIs/General-API-Information/Sumo-Logic-Endpoints-and-Firewall-Security
  endpoint: 'https://api.**.sumologic.com/api/v*/',

  // In case of request error, how much time to wait (in ms) before trying again
  retryDelay: 1000, // default

  // In case of request error, how many times to retry
  retryCount: 3, // default

  // the `timeZone` parameter of Create Search Job API, as described here:
  // https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API
  timezone: 'Asia/Tokyo', // default

  // the value `limit` parameter when paging through messages, as described here:
  // https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API
  searchPageLimit: 50000, // default

  // if we are waiting for the backend to gather more results, this is the delay between
  // checking if new results are available
  pollingDelay: 1000 // default
}
```

## Limitations

Aside from features that are simply not yet implemented:

-  When using in the browser, there is no way to identify certain backend errors: such as 401 or 403. This is due to limitations of Sumo Logic backend: they will not return CORS headers for such responses, and browser will block these requests due to same-origin policy violation. This means that we can not check response status in the code, and API call will simply return to you as "Network Error".

## TODOs

-  Implement aggregations and metrics
-  Improve test coverage

## License

ISC


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.2.0](https://github.com/moneytree/sumologic-client/compare/v0.1.0...v0.2.0) (2020-02-01)


### Features

* support records endpoint ([5964b67](https://github.com/moneytree/sumologic-client/commit/5964b67))



## [0.2.0](https://github.com/moneytree/sumologic-client/compare/v0.0.3...v0.2.0) (2019-08-19)


### Features

* support byReceiptTime parameter for search queries ([f986eb2](https://github.com/moneytree/sumologic-client/commit/f986eb2))

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.1.0](https://github.com/moneytree/sumologic-client/compare/v0.0.3...v0.1.0) (2019-08-19)


### Features

* support byReceiptTime parameter for search queries ([f986eb2](https://github.com/moneytree/sumologic-client/commit/f986eb2))



### [0.0.3](https://github.com/moneytree/sumologic-client/compare/v0.0.2...v0.0.3) (2019-07-18)


### Bug Fixes

* bump up documentation package ([86c565e](https://github.com/moneytree/sumologic-client/commit/86c565e))



### 0.0.2 (2019-07-12)


test/fixtures/00.js
module.exports = [{
  scope: 'https://api.jp.sumologic.com:443',
  method: 'POST',
  path: '/api/v1/search/jobs',
  body: {
    query: 'sample_query',
    from: '2019-06-25T10:14:31+09:00',
    to: '2019-06-25T17:14:31+09:00',
    timeZone: 'Asia/Tokyo'
  },
  status: 202,
  response: {
    id: '750D3ABE4460BA73',
    link: {
      rel: 'self',
      href: 'https://api.jp.sumologic.com/api/v1/search/jobs/750D3ABE4460BA73'
    }
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=Pxmydyk+9LMvAEfZH2h55jJve2ufmwdneWqzPcrul85wJo8SD0DoFmSBQIDhde3Sv1rp4Z/D3JXi7l0MLGtr0e1IhE6XhN8ihEAQoMPtEcPrJCRi7EYX0AJCS/03; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0z8u0k92qywg3ek8wy6br6dj2843.node0; Path=/api',
    'Location',
    'https://api.jp.sumologic.com/api/v1/search/jobs/750D3ABE4460BA73'
  ]
},
{
  scope: 'https://api.jp.sumologic.com:443',
  method: 'GET',
  path: '/api/v1/search/jobs/750D3ABE4460BA73',
  body: '',
  status: 200,
  response: {
    state: 'DONE GATHERING RESULTS',
    histogramBuckets: [
      {
        startTimestamp: 1561425300000,
        length: 300000,
        count: 576
      },
      {
        startTimestamp: 1561425271000,
        length: 29000,
        count: 78
      }
    ],
    messageCount: 3,
    recordCount: 0,
    pendingWarnings: [],
    pendingErrors: []
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=G6vGpwo8A/C1Zu+1IpQU3WVBpYJqr+JToBZ0+s8l3EPYaePY5gVHxz8IJUTa1Mf/YdJnS09GFLakCjPP3v4z5CAqyObDcQTWUb+UI8ddmyxFEuSt+C7A8xApOZlm; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0shrdo704xzmy1sgazlho1l8z42840.node0; Path=/api',
    'Vary',
    'Accept-Encoding, User-Agent'
  ]
}, {
  scope: 'https://api.jp.sumologic.com:443',
  method: 'GET',
  path: '/api/v1/search/jobs/750D3ABE4460BA73/messages?offset=0&limit=50000',
  body: '',
  status: 200,
  response: {
    fields: [
      {
        name: 'sample-field',
        fieldType: 'long',
        keyField: false
      }
    ],
    messages: [
      {
        map: {
          msg: 'message1'
        }
      },
      {
        map: {
          msg: 'message2'
        }
      },
      {
        map: {
          msg: 'message3'
        }
      }
    ]
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=TwtM7wDsw4K/kc52LbyjxJZNENNUeDU/t0/9NVDxUU/0EqLuCyor/Ym9eevsol1hl8BTFVnuVEIGxdkgM0mqzDEUHARpwhHjl8axm8NKxpJFYV+BMU9lyuLbMdXo; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node013d6vnkzzyjexr2dcnemtws1d2841.node0; Path=/api',
    'Vary',
    'Accept-Encoding, User-Agent'
  ]
}, {
  scope: 'https://api.jp.sumologic.com:443',
  method: 'DELETE',
  path: '/api/v1/search/jobs/750D3ABE4460BA73',
  body: '',
  status: 200,
  response: {
    id: '750D3ABE4460BA73'
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:37 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=lmcUMXglhcyc2AeVaWGWPZB9eWMxQ8VEJHBgZlUuqa37n5yLlvp01+7LL2Z8ghzQkm7CxmBlzW3u+Anhi5HwJIuMeJ/hDQFpM5Xw2dJNynVWe8sR0xO4QuR5yU4B; Expires=Wed, 17 Jul 2019 04:10:37 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0mkt2s73n3p8umrb4rofhd25a2842.node0; Path=/api'
  ]
}];


test/fixtures/01.js
module.exports = [{
  scope: 'https://api.jp.sumologic.com:443',
  method: 'POST',
  path: '/api/v1/search/jobs',
  body: {
    query: 'sample_query',
    from: '2019-06-25T10:14:31+09:00',
    to: '2019-06-25T17:14:31+09:00',
    timeZone: 'Asia/Tokyo',
    byReceiptTime: 'true'
  },
  status: 202,
  response: {
    id: '750D3ABE4460BA73',
    link: {
      rel: 'self',
      href: 'https://api.jp.sumologic.com/api/v1/search/jobs/750D3ABE4460BA73'
    }
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=Pxmydyk+9LMvAEfZH2h55jJve2ufmwdneWqzPcrul85wJo8SD0DoFmSBQIDhde3Sv1rp4Z/D3JXi7l0MLGtr0e1IhE6XhN8ihEAQoMPtEcPrJCRi7EYX0AJCS/03; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0z8u0k92qywg3ek8wy6br6dj2843.node0; Path=/api',
    'Location',
    'https://api.jp.sumologic.com/api/v1/search/jobs/750D3ABE4460BA73'
  ]
},
{
  scope: 'https://api.jp.sumologic.com:443',
  method: 'GET',
  path: '/api/v1/search/jobs/750D3ABE4460BA73',
  body: '',
  status: 200,
  response: {
    state: 'DONE GATHERING RESULTS',
    histogramBuckets: [
      {
        startTimestamp: 1561425300000,
        length: 300000,
        count: 576
      },
      {
        startTimestamp: 1561425271000,
        length: 29000,
        count: 78
      }
    ],
    messageCount: 3,
    recordCount: 0,
    pendingWarnings: [],
    pendingErrors: []
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=G6vGpwo8A/C1Zu+1IpQU3WVBpYJqr+JToBZ0+s8l3EPYaePY5gVHxz8IJUTa1Mf/YdJnS09GFLakCjPP3v4z5CAqyObDcQTWUb+UI8ddmyxFEuSt+C7A8xApOZlm; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0shrdo704xzmy1sgazlho1l8z42840.node0; Path=/api',
    'Vary',
    'Accept-Encoding, User-Agent'
  ]
}, {
  scope: 'https://api.jp.sumologic.com:443',
  method: 'GET',
  path: '/api/v1/search/jobs/750D3ABE4460BA73/messages?offset=0&limit=50000',
  body: '',
  status: 200,
  response: {
    fields: [
      {
        name: 'sample-field',
        fieldType: 'long',
        keyField: false
      }
    ],
    messages: [
      {
        map: {
          msg: 'message1'
        }
      },
      {
        map: {
          msg: 'message2'
        }
      },
      {
        map: {
          msg: 'message3'
        }
      }
    ]
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=TwtM7wDsw4K/kc52LbyjxJZNENNUeDU/t0/9NVDxUU/0EqLuCyor/Ym9eevsol1hl8BTFVnuVEIGxdkgM0mqzDEUHARpwhHjl8axm8NKxpJFYV+BMU9lyuLbMdXo; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node013d6vnkzzyjexr2dcnemtws1d2841.node0; Path=/api',
    'Vary',
    'Accept-Encoding, User-Agent'
  ]
}, {
  scope: 'https://api.jp.sumologic.com:443',
  method: 'DELETE',
  path: '/api/v1/search/jobs/750D3ABE4460BA73',
  body: '',
  status: 200,
  response: {
    id: '750D3ABE4460BA73'
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:37 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=lmcUMXglhcyc2AeVaWGWPZB9eWMxQ8VEJHBgZlUuqa37n5yLlvp01+7LL2Z8ghzQkm7CxmBlzW3u+Anhi5HwJIuMeJ/hDQFpM5Xw2dJNynVWe8sR0xO4QuR5yU4B; Expires=Wed, 17 Jul 2019 04:10:37 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0mkt2s73n3p8umrb4rofhd25a2842.node0; Path=/api'
  ]
}];


test/fixtures/02.js
module.exports = [{
  scope: 'https://api.jp.sumologic.com:443',
  method: 'POST',
  path: '/api/v1/search/jobs',
  body: {
    query: 'sample_query',
    from: '2019-06-25T10:14:31+09:00',
    to: '2019-06-25T17:14:31+09:00',
    timeZone: 'Asia/Tokyo'
  },
  status: 202,
  response: {
    id: '750D3ABE4460BA73',
    link: {
      rel: 'self',
      href: 'https://api.jp.sumologic.com/api/v1/search/jobs/750D3ABE4460BA73'
    }
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=Pxmydyk+9LMvAEfZH2h55jJve2ufmwdneWqzPcrul85wJo8SD0DoFmSBQIDhde3Sv1rp4Z/D3JXi7l0MLGtr0e1IhE6XhN8ihEAQoMPtEcPrJCRi7EYX0AJCS/03; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0z8u0k92qywg3ek8wy6br6dj2843.node0; Path=/api',
    'Location',
    'https://api.jp.sumologic.com/api/v1/search/jobs/750D3ABE4460BA73'
  ]
},
{
  scope: 'https://api.jp.sumologic.com:443',
  method: 'GET',
  path: '/api/v1/search/jobs/750D3ABE4460BA73',
  body: '',
  status: 200,
  response: {
    state: 'DONE GATHERING RESULTS',
    histogramBuckets: [
      {
        startTimestamp: 1561425300000,
        length: 300000,
        count: 576
      },
      {
        startTimestamp: 1561425271000,
        length: 29000,
        count: 78
      }
    ],
    messageCount: 3,
    recordCount: 3,
    pendingWarnings: [],
    pendingErrors: []
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=G6vGpwo8A/C1Zu+1IpQU3WVBpYJqr+JToBZ0+s8l3EPYaePY5gVHxz8IJUTa1Mf/YdJnS09GFLakCjPP3v4z5CAqyObDcQTWUb+UI8ddmyxFEuSt+C7A8xApOZlm; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0shrdo704xzmy1sgazlho1l8z42840.node0; Path=/api',
    'Vary',
    'Accept-Encoding, User-Agent'
  ]
}, {
  scope: 'https://api.jp.sumologic.com:443',
  method: 'GET',
  path: '/api/v1/search/jobs/750D3ABE4460BA73/records?offset=0&limit=50000',
  body: '',
  status: 200,
  response: {
    fields: [
      {
        name: 'sample-field',
        fieldType: 'long',
        keyField: false
      }
    ],
    records: [
      {
        "map": {
          "_count": "1",
          "key": "aggregation key1",
        }
      },
      {
        "map": {
          "_count": "2",
          "key": "aggregation key2",
        }
      },
      {
        "map": {
          "_count": "3",
          "key": "aggregation key3",
        }
      }
    ]
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:36 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=TwtM7wDsw4K/kc52LbyjxJZNENNUeDU/t0/9NVDxUU/0EqLuCyor/Ym9eevsol1hl8BTFVnuVEIGxdkgM0mqzDEUHARpwhHjl8axm8NKxpJFYV+BMU9lyuLbMdXo; Expires=Wed, 17 Jul 2019 04:10:36 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node013d6vnkzzyjexr2dcnemtws1d2841.node0; Path=/api',
    'Vary',
    'Accept-Encoding, User-Agent'
  ]
}, {
  scope: 'https://api.jp.sumologic.com:443',
  method: 'DELETE',
  path: '/api/v1/search/jobs/750D3ABE4460BA73',
  body: '',
  status: 200,
  response: {
    id: '750D3ABE4460BA73'
  },
  rawHeaders: [
    'Date',
    'Wed, 10 Jul 2019 04:10:37 GMT',
    'Content-Type',
    'application/json;charset=utf-8',
    'Transfer-Encoding',
    'chunked',
    'Connection',
    'close',
    'Set-Cookie',
    'AWSALB=lmcUMXglhcyc2AeVaWGWPZB9eWMxQ8VEJHBgZlUuqa37n5yLlvp01+7LL2Z8ghzQkm7CxmBlzW3u+Anhi5HwJIuMeJ/hDQFpM5Xw2dJNynVWe8sR0xO4QuR5yU4B; Expires=Wed, 17 Jul 2019 04:10:37 GMT; Path=/',
    'Cache-Control',
    'no-cache, no-store, max-age=0, must-revalidate',
    'Pragma',
    'no-cache',
    'Expires',
    'Thu, 01 Jan 1970 00:00:00 GMT',
    'X-XSS-Protection',
    '1; mode=block',
    'X-Frame-Options',
    'DENY',
    'X-Content-Type-Options',
    'nosniff',
    'Set-Cookie',
    'JSESSIONID=node0mkt2s73n3p8umrb4rofhd25a2842.node0; Path=/api'
  ]
}];


src/search.js
const axios = require('axios');
const moment = require('moment');
require('moment-timezone');
const debug = require('debug');

const log = debug('sumologic-search');

class Search {
  get config() {
    return this._config;
  }

  constructor(config) {
    if (!config.hasOwnProperty('accessId')) {
      throw new Error('Please pass in accessId');
    }

    if (!config.hasOwnProperty('accessKey')) {
      throw new Error('Please pass in accessKey');
    }

    if (!config.hasOwnProperty('endpoint')) {
      throw new Error('Please pass in endpoint');
    }

    config.retryDelay = config.retryDelay || Search.CFG_DEFAULT_RETRY_DELAY;
    config.retryCount = config.retryCount || Search.CFG_DEFAULT_RETRY_COUNT;
    config.timezone = config.timezone || Search.CFG_DEFAULT_TIMEZONE;
    config.searchPageLimit = config.searchPageLimit || Search.CFG_DEFAULT_PAGE_LIMIT;
    config.pollingDelay = config.pollingDelay || Search.CFG_DEFAULT_POLLING_DELAY;

    this._config = config;
  }

  async request(params) {
    const headers = params.headers || {};
    headers['Content-Type'] = 'application/json';
    headers.Authorization = `Basic ${this._getBasicAuthSecret()}`;

    params.headers = headers;
    params.baseURL = this.config.endpoint;
    params.validateStatus = () => true;
    params.transformRequest = (data) => JSON.stringify(data);

    let remainingRetries = this.config.retryCount;

    log(`requesting ${JSON.stringify(params)}`);

    while (true) {
      let response;

      try {
        // TODO
        // use axios.create
        response = await axios.request(params);
      } catch (e) {
        log(`request error: ${e.message}`);
        if (!e.status && remainingRetries > 0) {
          log('network error, retrying');
          remainingRetries--;
          await Search._wait(this.config.retryDelay);
          continue;
        } else {
          log('ran out of retry attempts');
          throw e;
        }
      }

      log(`response code=${response.status}`);

      // NOTE
      // Error handling won't work properly in the browser because they
      // do not include CORS headers in case of non-2xx responses.
      if (response.status === 429 || response.status === 504) {
        // rate limit exceeded, or gateway timeout: always retry
        log('got 429/504, retrying');
        await Search._wait(this.config.retryDelay);
        continue;
      }

      if (response.status === 500 || response.status === 503) {
        // backend error, or service unavailable: retry
        if (remainingRetries > 0) {
          log('got 500/503, retrying');
          await Search._wait(this.config.retryDelay);
          remainingRetries--;
          continue;
        } else {
          log('ran out of retry attempts');
          throw new Error('Exceeded retry count for 500/503 errors.');
        }
      }

      if (response.status >= 400) {
        log(`got ${response.status} from server, failing immediately`);
        const error = new Error('A 4xx response from Sumo Logic API. Debug the' +
                                'attached `response` object for reference.');
        error.response = response;
        throw error;
      }

      return response;
    }
  }

  async newSearchJob(_queryParams) {
    log('newSearchJob');

    const queryParams = {
      query: _queryParams.query
    };

    queryParams.from = moment(_queryParams.from).tz(this.config.timezone).format();
    queryParams.to = moment(_queryParams.to).tz(this.config.timezone).format();
    queryParams.timeZone = _queryParams.timeZone || this.config.timezone;

    if (_queryParams.byReceiptTime === true || _queryParams.byReceiptTime === 'true') {
      queryParams.byReceiptTime = 'true';
    }

    const requestParams = {
      method: 'post',
      url: '/search/jobs',
      data: queryParams
    };

    const response = await this.request(requestParams);

    return response.data.id;
  }

  async getJobState(id) {
    log('getJobState');

    const requestParams = {
      method: 'get',
      url: `/search/jobs/${id}`
    };

    const response = await this.request(requestParams);

    return response.data;
  }

  /**
   * Gets results gathered so far, from a given offset. This does not take the current job state
   * into consideration. E.g. this method may be called when results are still being gathered, in
   * which case this method will return results gathered **so far**. If you call this while the
   * results are still being gathered, be sure to call this again with the corresponding offset.
   */
  async getJobResults(type, id, offset = 0) {
    log('getJobResults');

    const requestParams = {
      method: 'get',
      url: `/search/jobs/${id}/${type}s?offset=${offset}&limit=${this.config.searchPageLimit}`
    };

    const response = await this.request(requestParams);

    return response.data;
  }

  async deleteSearchJob(id) {
    log('deleteSearchJob');

    const requestParams = {
      method: 'delete',
      url: `/search/jobs/${id}`
    };
    await this.request(requestParams);
  }

  cancel() {
    // eslint-disable-next-line no-new-object
    this._nonce = new Object();
  }

  async *_getIterator(type, queryParams) {
    const jobId = await this.newSearchJob(queryParams);
    let myNonce;

    // eslint-disable-next-line no-new-object
    this._nonce = myNonce = new Object();

    let resultsGatheredSoFar = 0;
    let state = null;
    const countAttribute = `${type}Count`;

    while (true) {
      if (this._nonce !== myNonce) {
        log('another operation has been started, canceling search');
        await this.deleteSearchJob(jobId);
        throw new Search.ERROR_SEARCH_CANCELED();
      }

      // refresh the state, unless already done gathering results:
      if (!(state && state.state === Search.STATE_DONE_GATHERING_RESULTS)) {
        state = await this.getJobState(jobId);
        log(`state=${state.state}, cnt=${state[countAttribute]}`);
      }

      if (state.state === Search.STATE_NOT_STARTED) {
        // not started yet, wait for a while
        log('job not started yet');
        await Search._wait(this.config.pollingDelay);
        continue;
      }

      if (state.state === Search.STATE_CANCELED) {
        log('job has been canceled');
        throw new Error('Search job has been canceled unexpectedly.');
      }

      if (state.state === Search.STATE_FORCE_PAUSED) {
        log('job has been force paused');
        throw new Error('Search job has been force paused. This behavior is currently not supported.');
      }

      const results = await this.getJobResults(type, jobId, resultsGatheredSoFar);

      yield { state, results };

      resultsGatheredSoFar += results[`${type}s`].length;
      log(`gathered so far: ${resultsGatheredSoFar}`);

      if (state.state === Search.STATE_DONE_GATHERING_RESULTS) {
        if (state[countAttribute] === resultsGatheredSoFar) {
          log('job done gathering results, and we downloaded everything; wrapping up');
          await this.deleteSearchJob(jobId);
          return;
        }
        log('job done gathering results, but we are still downloading them');
      } else {
        log('sumologic is still gathering results, waiting a bit and polling again');
        await Search._wait(this.config.pollingDelay);
      }
    }
  }


  /**
   * Gets message results gathered so far, from a given offset. This does not take the current job state
   * into consideration. E.g. this method may be called when results are still being gathered, in
   * which case this method will return results gathered **so far**. If you call this while the
   * results are still being gathered, be sure to call this again with the corresponding offset.
   */
  async getMessageIterator(id, offset = 0) {
    return await this._getIterator('message', id, offset);
  }

  /**
   * Gets record results gathered so far, from a given offset. This does not take the current job state
   * into consideration. E.g. this method may be called when results are still being gathered, in
   * which case this method will return results gathered **so far**. If you call this while the
   * results are still being gathered, be sure to call this again with the corresponding offset.
   */
  async getRecordIterator(id, offset = 0) {
    return await this._getIterator('record', id, offset);
  }

  _getBasicAuthSecret() {
    // eslint-disable-next-line no-undef
    let btoa = typeof window === 'undefined' ? undefined : window.btoa;
    if (typeof Buffer !== 'undefined') {
      btoa = (str) => Buffer.from(str, 'utf8').toString('base64');
    }

    const secret = btoa(`${this.config.accessId}:${this.config.accessKey}`);

    return secret;
  }

  static _wait(waitms) {
    return new Promise((resolve) => {
      setTimeout(resolve, waitms);
    });
  }
}

// default configuration keys:
Search.CFG_DEFAULT_RETRY_DELAY = 1000;
Search.CFG_DEFAULT_RETRY_COUNT = 3;
Search.CFG_DEFAULT_TIMEZONE = 'Asia/Tokyo';
Search.CFG_DEFAULT_POLLING_DELAY = 1000;
Search.CFG_DEFAULT_PAGE_LIMIT = 50000;

// search job states;
Search.STATE_GATHERING_RESULTS = 'GATHERING RESULTS';
// Search job has not been started yet.
Search.STATE_NOT_STARTED = 'NOT STARTED';
// Query that is paused by the system. It is true only for non-aggregate queries that are paused at the limit of 100k.
// This limit is dynamic and may vary from customer to customer.
Search.STATE_FORCE_PAUSED = 'FORCE PAUSED';
Search.STATE_DONE_GATHERING_RESULTS = 'DONE GATHERING RESULTS';
// The search job has been canceled. Note the spelling has two L letters.
Search.STATE_CANCELED = 'CANCELLED';

// endpoints
Search.ENDPOINT_JP = 'https://api.jp.sumologic.com/api/v1/';

// errors
Search.ERROR_SEARCH_CANCELED = class SearchCanceledError extends Error {
  constructor() {
    super('Search has been canceled.');
    this.name = this.constructor.name;
    Error.captureStackTrace(this, this.constructor);
  }
};

module.exports = Search;
